File: internal/translator/response/openai-to-claude.go
package response

import (
	"encoding/json"
	"flamerouter/internal/translator"
	"flamerouter/internal/translator/concerns"
	"flamerouter/internal/translator/schema"
	"strconv"
	"strings"
	"time"
)

// ClaudeOAuthToolPrefix is the prefix used for Claude OAuth tools.
const ClaudeOAuthToolPrefix = "proxy_"

func init() {
	translator.Register(translator.FormatOpenAI, translator.FormatClaude, nil, openaiToClaudeResponse)
}

func parseOpenAIUsage(chunk map[string]any, state *concerns.ResponseState) {
	chunkUsage, ok := chunk["usage"].(map[string]any)
	if !ok {
		return
	}

	promptTokens := getInt(chunkUsage, "prompt_tokens")
	outputTokens := getInt(chunkUsage, "completion_tokens")
	cacheRead := 0
	cacheCreate := 0

	if details, ok := chunkUsage["prompt_tokens_details"].(map[string]any); ok {
		cacheRead = getInt(details, "cached_tokens")
		cacheCreate = getInt(details, "cache_creation_tokens")
	}

	inputTokens := promptTokens - cacheRead - cacheCreate
	state.Usage = &concerns.UsageInfo{
		PromptTokens:             promptTokens,
		CompletionTokens:         outputTokens,
		TotalTokens:              promptTokens + outputTokens,
		InputTokens:              inputTokens,
		OutputTokens:             outputTokens,
		CacheReadInputTokens:     cacheRead,
		CacheCreationInputTokens: cacheCreate,
		CacheReadTokens:          0,
		CacheCreateTokens:        0,
	}
}

func resolveClaudeMessageID(chunk map[string]any, currentID string) string {
	msgID := currentID
	if id, idOk := chunk["id"].(string); idOk {
		msgID = id
	}

	msgID = strings.TrimPrefix(msgID, "chatcmpl-")

	if msgID == "" || msgID == "chat" || len(msgID) < 8 {
		if ext, ok := chunk["extend_fields"].(map[string]any); ok {
			if rid, ok := ext["requestId"].(string); ok && rid != "" {
				msgID = rid
			}
		}
	}

	if msgID == "" {
		msgID = "msg_" + strconv.FormatInt(time.Now().UnixMilli(), 10)
	}

	return msgID
}

func handleOpenAIToClaudeMessageStart(chunk map[string]any, state *concerns.ResponseState) []map[string]any {
	if state.MessageStartSent {
		return nil
	}

	state.MessageStartSent = true
	state.MessageID = resolveClaudeMessageID(chunk, state.MessageID)

	if model, modelOk := chunk["model"].(string); modelOk {
		state.Model = model
	}

	if state.Model == "" {
		state.Model = schema.ModelFallback
	}

	state.NextBlockIndex = 0

	return []map[string]any{
		{
			"type": "message_start",
			"message": map[string]any{
				"id":            state.MessageID,
				"type":          "message",
				"role":          schema.RoleAssistant,
				"model":         state.Model,
				"content":       []any{},
				"stop_reason":   nil,
				"stop_sequence": nil,
				"usage": map[string]any{
					"input_tokens":  0,
					"output_tokens": 0,
				},
			},
		},
	}
}

func handleClaudeReasoning(delta map[string]any, state *concerns.ResponseState) []map[string]any {
	reasoningContent := concerns.ExtractReasoningText(delta)
	if reasoningContent == "" {
		return nil
	}

	var results []map[string]any

	if state.TextBlockStarted && !state.TextBlockClosed {
		results = append(results, map[string]any{
			"type":  "content_block_stop",
			"index": state.TextBlockIndex,
		})
		state.TextBlockStarted = false
		state.TextBlockClosed = true
	}

	if !state.ThinkingBlockStarted {
		state.ThinkingBlockIndex = state.NextBlockIndex
		state.NextBlockIndex++
		state.ThinkingBlockStarted = true
		results = append(results, map[string]any{
			"type":  "content_block_start",
			"index": state.ThinkingBlockIndex,
			"content_block": map[string]any{
				"type":     schema.ClaudeBlockThinking,
				"thinking": "",
			},
		})
	}

	results = append(results, map[string]any{
		"type":  "content_block_delta",
		"index": state.ThinkingBlockIndex,
		"delta": map[string]any{
			"type":     "thinking_delta",
			"thinking": reasoningContent,
		},
	})

	return results
}

func handleClaudeContent(delta map[string]any, state *concerns.ResponseState) []map[string]any {
	content, ok := delta["content"].(string)
	if !ok || content == "" {
		return nil
	}

	var results []map[string]any

	if state.ThinkingBlockStarted {
		results = append(results, map[string]any{
			"type":  "content_block_stop",
			"index": state.ThinkingBlockIndex,
		})
		state.ThinkingBlockStarted = false
	}

	if !state.TextBlockStarted {
		state.TextBlockIndex = state.NextBlockIndex
		state.NextBlockIndex++
		state.TextBlockStarted = true
		state.TextBlockClosed = false
		results = append(results, map[string]any{
			"type":  "content_block_start",
			"index": state.TextBlockIndex,
			"content_block": map[string]any{
				"type": schema.ClaudeBlockText,
				"text": "",
			},
		})
	}

	results = append(results, map[string]any{
		"type":  "content_block_delta",
		"index": state.TextBlockIndex,
		"delta": map[string]any{
			"type": "text_delta",
			"text": content,
		},
	})

	return results
}

func handleClaudeToolCallStart(tc map[string]any, idx int, id string, state *concerns.ResponseState) []map[string]any {
	var results []map[string]any

	if state.ThinkingBlockStarted {
		results = append(results, map[string]any{
			"type":  "content_block_stop",
			"index": state.ThinkingBlockIndex,
		})
		state.ThinkingBlockStarted = false
	}

	if state.TextBlockStarted && !state.TextBlockClosed {
		results = append(results, map[string]any{
			"type":  "content_block_stop",
			"index": state.TextBlockIndex,
		})
		state.TextBlockStarted = false
		state.TextBlockClosed = true
	}

	name := ""

	if fn, ok := tc["function"].(map[string]any); ok && fn != nil {
		if n, nOk := fn["name"].(string); nOk {
			name = n
		}
	}

	state.ToolCalls[idx] = &concerns.ToolCallInfo{
		ID:         id,
		Name:       name,
		BlockIndex: state.NextBlockIndex,
	}
	state.NextBlockIndex++

	toolName := name
	if mapped, ok := state.ToolNameMap[name]; ok {
		toolName = mapped
	}

	toolName = strings.TrimPrefix(toolName, ClaudeOAuthToolPrefix)

	results = append(results, map[string]any{
		"type":  "content_block_start",
		"index": state.ToolCalls[idx].BlockIndex,
		"content_block": map[string]any{
			"type":  schema.ClaudeBlockToolUse,
			"id":    id,
			"name":  toolName,
			"input": map[string]any{},
		},
	})

	return results
}

func processSingleClaudeToolCall(tc map[string]any, state *concerns.ResponseState) []map[string]any {
	idx := 0
	if v, vOk := tc["index"].(float64); vOk {
		idx = int(v)
	}

	var results []map[string]any

	if id, idOk := tc["id"].(string); idOk && id != "" {
		if _, exists := state.ToolCalls[idx]; !exists {
			results = append(results, handleClaudeToolCallStart(tc, idx, id, state)...)
		}
	}

	if fn, fnOk := tc["function"].(map[string]any); fnOk {
		if args, argsOk := fn["arguments"].(string); argsOk && args != "" {
			if _, exists := state.ToolCalls[idx]; exists {
				state.ToolArgBuffers[idx] += args
			}
		}
	}

	return results
}

func handleClaudeToolCalls(toolCalls []any, state *concerns.ResponseState) []map[string]any {
	results := make([]map[string]any, 0, len(toolCalls))

	for _, tcRaw := range toolCalls {
		tc, ok := tcRaw.(map[string]any)
		if !ok || tc == nil {
			continue
		}

		results = append(results, processSingleClaudeToolCall(tc, state)...)
	}

	return results
}

func closeClaudeContentBlocks(state *concerns.ResponseState) []map[string]any {
	results := make([]map[string]any, 0, len(state.ToolCalls)+2)

	if state.ThinkingBlockStarted {
		results = append(results, map[string]any{
			"type":  "content_block_stop",
			"index": state.ThinkingBlockIndex,
		})
		state.ThinkingBlockStarted = false
	}

	if state.TextBlockStarted && !state.TextBlockClosed {
		results = append(results, map[string]any{
			"type":  "content_block_stop",
			"index": state.TextBlockIndex,
		})
		state.TextBlockStarted = false
		state.TextBlockClosed = true
	}

	for _, toolInfo := range state.ToolCalls {
		buffered := state.ToolArgBuffers[toolInfo.BlockIndex]
		if buffered != "" {
			sanitized := sanitizeToolArgs(toolInfo.Name, buffered)
			results = append(results, map[string]any{
				"type":  "content_block_delta",
				"index": toolInfo.BlockIndex,
				"delta": map[string]any{
					"type":         "input_json_delta",
					"partial_json": sanitized,
				},
			})
		}

		results = append(results, map[string]any{
			"type":  "content_block_stop",
			"index": toolInfo.BlockIndex,
		})
	}

	return results
}

func handleClaudeFinishReason(fr string, state *concerns.ResponseState) []map[string]any {
	results := closeClaudeContentBlocks(state)

	state.FinishReason = fr
	stopReason := convertOpenAIToClaudeFinish(fr)

	finalUsage := state.Usage
	if finalUsage == nil {
		finalUsage = &concerns.UsageInfo{
			PromptTokens:             0,
			CompletionTokens:         0,
			TotalTokens:              0,
			InputTokens:              0,
			OutputTokens:             0,
			CacheReadTokens:          0,
			CacheCreateTokens:        0,
			CacheReadInputTokens:     0,
			CacheCreationInputTokens: 0,
		}
	}

	usageMap := map[string]any{
		"input_tokens":  finalUsage.InputTokens,
		"output_tokens": finalUsage.OutputTokens,
	}
	if finalUsage.CacheReadInputTokens > 0 {
		usageMap["cache_read_input_tokens"] = finalUsage.CacheReadInputTokens
	}

	if finalUsage.CacheCreationInputTokens > 0 {
		usageMap["cache_creation_input_tokens"] = finalUsage.CacheCreationInputTokens
	}

	results = append(results, map[string]any{
		"type": "message_delta",
		"delta": map[string]any{
			"stop_reason": stopReason,
		},
		"usage": usageMap,
	})
	results = append(results, map[string]any{
		"type": "message_stop",
	})

	return results
}

func extractOpenAIChoice(chunk map[string]any) (map[string]any, map[string]any, bool) {
	choices, ok := chunk["choices"].([]any)
	if !ok || len(choices) == 0 {
		return nil, nil, false
	}

	choice, ok := choices[0].(map[string]any)
	if !ok || choice == nil {
		return nil, nil, false
	}

	delta, ok := choice["delta"].(map[string]any)
	if !ok || delta == nil {
		return nil, nil, false
	}

	return choice, delta, true
}

func openaiToClaudeResponse(chunk map[string]any, state *concerns.ResponseState) []map[string]any {
	if chunk == nil {
		return nil
	}

	choice, delta, ok := extractOpenAIChoice(chunk)
	if !ok {
		return nil
	}

	parseOpenAIUsage(chunk, state)

	results := handleOpenAIToClaudeMessageStart(chunk, state)
	results = append(results, handleClaudeReasoning(delta, state)...)
	results = append(results, handleClaudeContent(delta, state)...)

	if toolCalls, tcOk := delta["tool_calls"].([]any); tcOk {
		results = append(results, handleClaudeToolCalls(toolCalls, state)...)
	}

	if fr, frOk := choice["finish_reason"].(string); frOk && fr != "" {
		results = append(results, handleClaudeFinishReason(fr, state)...)
	}

	if len(results) == 0 {
		return nil
	}

	return results
}

func sanitizeToolArgs(toolName, argsJSON string) string {
	var args map[string]any
	if err := json.Unmarshal([]byte(argsJSON), &args); err != nil {
		return argsJSON
	}

	name := strings.TrimPrefix(toolName, ClaudeOAuthToolPrefix)
	if name == "Read" {
		sanitizeReadArgs(args)
	}

	result, err := json.Marshal(args)
	if err != nil {
		return argsJSON
	}

	return string(result)
}

func sanitizeReadLimit(args map[string]any) {
	limit, ok := args["limit"]
	if !ok {
		return
	}

	switch v := limit.(type) {
	case string:
		if parsed, err := strconv.Atoi(v); err == nil {
			args["limit"] = parsed
		}
	case float64:
		if v > 2000 {
			args["limit"] = float64(2000)
		}

		if v < 1 {
			delete(args, "limit")
		}
	}
}

func sanitizeReadOffset(args map[string]any) {
	offset, ok := args["offset"]
	if !ok {
		return
	}

	switch v := offset.(type) {
	case string:
		if parsed, err := strconv.Atoi(v); err == nil {
			args["offset"] = parsed
		}
	case float64:
		if v < 0 {
			args["offset"] = float64(0)
		}
	}
}

func sanitizeReadArgs(args map[string]any) {
	sanitizeReadLimit(args)
	sanitizeReadOffset(args)

	if pages, ok := args["pages"]; ok {
		filePath := ""
		if fp, fpOk := args["file_path"].(string); fpOk {
			filePath = fp
		}

		if !isValidPdfPagesArg(filePath, pages) {
			delete(args, "pages")
		}
	}
}

func isValidPdfPagesArg(filePath string, pages any) bool {
	if filePath == "" || len(filePath) < 4 {
		return false
	}

	if strings.ToLower(filePath[len(filePath)-4:]) != ".pdf" {
		return false
	}

	pagesStr, ok := pages.(string)
	if !ok {
		return false
	}

	for _, c := range pagesStr {
		if (c < '0' || c > '9') && c != '-' {
			return false
		}
	}

	return true
}

func convertOpenAIToClaudeFinish(reason string) string {
	switch reason {
	case "stop":
		return schema.ClaudeStopEndTurn
	case "length":
		return schema.ClaudeStopMaxTokens
	case "tool_calls":
		return schema.ClaudeStopToolUse
	case "content_filter":
		return schema.ClaudeStopEndTurn
	default:
		return schema.ClaudeStopEndTurn
	}
}

func getInt(m map[string]any, key string) int {
	if m == nil {
		return 0
	}

	v, ok := m[key]
	if !ok {
		return 0
	}

	switch n := v.(type) {
	case float64:
		return int(n)
	case int:
		return n
	case int64:
		return int(n)
	default:
		return 0
	}
}

File: open-sse/src/translator/response/openai_to_claude.rs
//! openai SSE chunk → claude SSE chunk.
//! Port of open-sse/translator/response/openai-to-claude.js

use serde_json::{Value, json};

fn msg_id(state: &Value) -> String {
    state
        .get("messageId")
        .and_then(|m| m.as_str())
        .map(|s| s.to_string())
        .unwrap_or_else(|| {
            format!(
                "msg_{}",
                std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .map(|d| d.as_millis())
                    .unwrap_or(0)
            )
        })
}

fn model_str(state: &Value) -> String {
    state
        .get("model")
        .and_then(|m| m.as_str())
        .unwrap_or("")
        .to_string()
}

fn ensure_started(state: &mut Value, chunk: &Value) -> Vec<Value> {
    let mut out = Vec::new();
    let started = state
        .get("started")
        .and_then(|b| b.as_bool())
        .unwrap_or(false);
    if started {
        return out;
    }
    // capture id/model from first chunk
    if let Some(id) = chunk.get("id").and_then(|i| i.as_str()) {
        // openai gives "chatcmpl-..." — keep as-is (strip prefix optional)
        state["messageId"] = json!(id);
    }
    if let Some(m) = chunk.get("model") {
        state["model"] = m.clone();
    }
    out.push(json!({
        "type": "message_start",
        "message": {
            "id": msg_id(state),
            "type": "message",
            "role": "assistant",
            "model": model_str(state),
            "content": [],
            "stop_reason": null,
            "stop_sequence": null,
            "usage": {"input_tokens": 0, "output_tokens": 0}
        }
    }));
    state["started"] = json!(true);
    state["contentBlockIndex"] = json!(0);
    state["textBlockStarted"] = json!(false);
    out
}

fn from_openai_finish_claude(reason: &str) -> &'static str {
    match reason {
        "stop" => "end_turn",
        "length" => "max_tokens",
        "tool_calls" => "tool_use",
        _ => "end_turn",
    }
}

pub fn translate(chunk: Value, state: &mut Value) -> Option<Vec<Value>> {
    let mut results = Vec::new();
    results.extend(ensure_started(state, &chunk));

    let choices = chunk.get("choices").and_then(|c| c.as_array())?;
    let choice = choices.first()?;
    let delta = choice.get("delta").cloned().unwrap_or(json!({}));
    let finish_reason = choice.get("finish_reason").and_then(|f| f.as_str());

    // text content delta
    if let Some(content) = delta.get("content").and_then(|c| c.as_str())
        && !content.is_empty() {
            let started = state
                .get("textBlockStarted")
                .and_then(|b| b.as_bool())
                .unwrap_or(false);
            if !started {
                let idx = state
                    .get("contentBlockIndex")
                    .and_then(|i| i.as_u64())
                    .unwrap_or(0);
                results.push(json!({
                    "type": "content_block_start",
                    "index": idx,
                    "content_block": {"type":"text","text":""}
                }));
                state["textBlockStarted"] = json!(true);
            }
            let idx = state
                .get("contentBlockIndex")
                .and_then(|i| i.as_u64())
                .unwrap_or(0);
            results.push(json!({
                "type": "content_block_delta",
                "index": idx,
                "delta": {"type":"text_delta","text": content}
            }));
        }

    // reasoning_content → thinking block
    if let Some(reasoning) = delta.get("reasoning_content").and_then(|c| c.as_str())
        && !reasoning.is_empty() {
            let thinking_started = state
                .get("thinkingBlockStarted")
                .and_then(|b| b.as_bool())
                .unwrap_or(false);
            if !thinking_started {
                let idx = state
                    .get("contentBlockIndex")
                    .and_then(|i| i.as_u64())
                    .unwrap_or(0);
                results.push(json!({
                    "type": "content_block_start",
                    "index": idx,
                    "content_block": {"type":"thinking","thinking":""}
                }));
                state["thinkingBlockStarted"] = json!(true);
            }
            let idx = state
                .get("contentBlockIndex")
                .and_then(|i| i.as_u64())
                .unwrap_or(0);
            results.push(json!({
                "type": "content_block_delta",
                "index": idx,
                "delta": {"type":"thinking_delta","thinking": reasoning}
            }));
        }

    // tool_calls
    if let Some(tcs) = delta.get("tool_calls").and_then(|t| t.as_array()) {
        for tc in tcs {
            let idx = tc.get("index").and_then(|i| i.as_u64()).unwrap_or(0);
            let fn_delta = tc.get("function").cloned().unwrap_or(json!({}));
            // start of new tool call (has name)
            if let Some(name) = fn_delta.get("name").and_then(|n| n.as_str()) {
                let block_idx = state
                    .get("contentBlockIndex")
                    .and_then(|i| i.as_u64())
                    .unwrap_or(0)
                    + 1
                    + idx;
                let id = tc.get("id").and_then(|i| i.as_str()).unwrap_or("");
                results.push(json!({
                    "type":"content_block_start",
                    "index": block_idx,
                    "content_block": {
                        "type":"tool_use",
                        "id": id,
                        "name": name,
                        "input": {}
                    }
                }));
                if !state.get("toolCalls").is_some() {
                    state["toolCalls"] = json!({});
                }
                state["toolCalls"][idx.to_string()] =
                    json!({"blockIndex": block_idx, "id": id, "name": name});
            }
            if let Some(args) = fn_delta.get("arguments").and_then(|a| a.as_str())
                && !args.is_empty() {
                    let block_idx = state
                        .get("toolCalls")
                        .and_then(|t| t.get(idx.to_string()))
                        .and_then(|t| t.get("blockIndex"))
                        .and_then(|b| b.as_u64())
                        .unwrap_or(0);
                    results.push(json!({
                        "type":"content_block_delta",
                        "index": block_idx,
                        "delta": {"type":"input_json_delta","partial_json": args}
                    }));
                }
        }
    }

    // finish_reason
    if let Some(fr) = finish_reason {
        // close any open text/thinking blocks
        let text_open = state
            .get("textBlockStarted")
            .and_then(|b| b.as_bool())
            .unwrap_or(false);
        let thinking_open = state
            .get("thinkingBlockStarted")
            .and_then(|b| b.as_bool())
            .unwrap_or(false);
        let idx = state
            .get("contentBlockIndex")
            .and_then(|i| i.as_u64())
            .unwrap_or(0);
        if text_open || thinking_open {
            results.push(json!({"type":"content_block_stop","index": idx}));
        }

        let stop_reason = from_openai_finish_claude(fr);
        let usage = chunk.get("usage").cloned().unwrap_or(json!({}));
        let input_tokens = usage
            .get("prompt_tokens")
            .and_then(|v| v.as_u64())
            .unwrap_or(0);
        let output_tokens = usage
            .get("completion_tokens")
            .and_then(|v| v.as_u64())
            .unwrap_or(0);
        results.push(json!({
            "type":"message_delta",
            "delta": {"stop_reason": stop_reason, "stop_sequence": null},
            "usage": {"input_tokens": input_tokens, "output_tokens": output_tokens}
        }));
        results.push(json!({"type":"message_stop"}));
    } else if let Some(usage) = chunk.get("usage") {
        // usage-only chunk (rare) — merge into state for later
        state["pendingUsage"] = usage.clone();
    }

    if results.is_empty() {
        None
    } else {
        Some(results)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn first_chunk_emits_message_start() {
        let chunk = json!({
            "id":"chatcmpl-1","model":"gpt-4","object":"chat.completion.chunk",
            "choices":[{"index":0,"delta":{"role":"assistant"},"finish_reason":null}]
        });
        let mut state = json!({});
        let out = translate(chunk, &mut state).unwrap();
        assert_eq!(out[0]["type"], "message_start");
        assert_eq!(out[0]["message"]["id"], "chatcmpl-1");
    }

    #[test]
    fn text_delta_emits_block_start_then_delta() {
        let mut state = json!({});
        let c1 = json!({"id":"chatcmpl-1","model":"gpt-4","choices":[{"index":0,"delta":{"role":"assistant"},"finish_reason":null}]});
        translate(c1, &mut state);
        let c2 = json!({"id":"chatcmpl-1","model":"gpt-4","choices":[{"index":0,"delta":{"content":"hi"},"finish_reason":null}]});
        let out = translate(c2, &mut state).unwrap();
        let types: Vec<&str> = out.iter().map(|e| e["type"].as_str().unwrap()).collect();
        assert!(types.contains(&"content_block_start"));
        assert!(types.contains(&"content_block_delta"));
        let delta_ev = out
            .iter()
            .find(|e| e["type"] == "content_block_delta")
            .unwrap();
        assert_eq!(delta_ev["delta"]["text"], "hi");
    }

    #[test]
    fn finish_reason_stop_emits_message_delta_and_stop() {
        let mut state = json!({});
        let c1 = json!({"id":"c","model":"m","choices":[{"index":0,"delta":{"role":"assistant"},"finish_reason":null}]});
        translate(c1, &mut state);
        let c2 = json!({"id":"c","model":"m","choices":[{"index":0,"delta":{},"finish_reason":"stop"}],"usage":{"prompt_tokens":10,"completion_tokens":5}});
        let out = translate(c2, &mut state).unwrap();
        let types: Vec<&str> = out.iter().map(|e| e["type"].as_str().unwrap()).collect();
        assert!(types.contains(&"message_delta"));
        assert!(types.contains(&"message_stop"));
        let md = out.iter().find(|e| e["type"] == "message_delta").unwrap();
        assert_eq!(md["delta"]["stop_reason"], "end_turn");
        assert_eq!(md["usage"]["output_tokens"], 5);
    }

    #[test]
    fn finish_reason_length_maps_to_max_tokens() {
        let mut state = json!({});
        let c1 = json!({"id":"c","model":"m","choices":[{"index":0,"delta":{"role":"assistant"},"finish_reason":null}]});
        translate(c1, &mut state);
        let c2 = json!({"id":"c","model":"m","choices":[{"index":0,"delta":{},"finish_reason":"length"}]});
        let out = translate(c2, &mut state).unwrap();
        let md = out.iter().find(|e| e["type"] == "message_delta").unwrap();
        assert_eq!(md["delta"]["stop_reason"], "max_tokens");
    }
}
